Change the API boundary between camera and stream with initial improvement for nest expiring stream urls

allenporter opened the pull request:
## Breaking change

Removes the internal `stream` components `record` service.   This service is meant to be used internally only by `camera` so this is not expected to be a problem in practice as users should be using `camera.record` instead.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Change the API boundary between `camera` and `stream`, so that `camera` manages more of the lifecycle of `Stream`.  This is one of multiple improvements discussed in https://github.com/home-assistant/architecture/issues/482

The motivation is to address problems in https://github.com/home-assistant/core/issues/42793 where a stream worker is started, and we don't have a way to update with a stream url when the active url changes.  The first step here is to simplify
stop and restart the stream worker here, but there will be additional improvements to simplify `stream`, for example, by unifying `keepalive` logic within the worker and ensuring the HLS view can work across stream url changes.

This is a walkthrough of the changes in this PR with motivations to make it easier to review:
- Add a method in `stream` called `create_stream` that creates a `Stream` class which is now managed by `camera`
- Camera creates a single `Stream` per camera entity so that one decoder is active for a stream URL.  This moves the `dict` of stream url to `Stream` effectively into an instance method on the `Camera`.  (Caveat: No longer dedups if multiple cameras share a `stream_source`, which doesn't seem worth handling).  Stream still manages idle timeout and shutdown of streams.  Add documentation explaining the lifecycle of the stream.
- Add a `endpoint_url` updating the access token
- Share code in `camera` for creating a stream, (e.g. unify logic for `keepalive` within `camera`)
- Remove the record service, which is now a method on `Stream`
- `nest` shuts down a `Stream` when the url changes. (Note: this may mean ATTR_STREAMS grows, but intent was to keep this PR small -- may want advice here, or a fast follow fix is OK)
- Tests are updated to use new API, e.g. `preload_stream` is replaced with the new `create_stream` API
- All test coverage from `stream/test_init.py` for record has moved to `stream/test_recorder.py`.


Future improvements to follow:
- Nest should update the existing `Stream` with the new URL (left out to keep PR small)
- Unify stream url updates and keepalive within `Stream`
- Change relationship between `Stream` and `StreamOutput` and `stream_worker` to ensure continuity at the HTS view layer (gold!) and expose fewer "internal" methods on `Stream` now that it is used in `Camera`.

Looking for feedback specifically on:
- Is making `Stream` the public API the right direction to move? (future PRs to remove other methods on it)
- Does this split of lifecycle make sense
- The discussion in https://github.com/home-assistant/architecture/issues/482 recommended creating a `create_stream_worker` but I thought that focusing on `Stream` as the right API made sense rather than introducing a new one at this point. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://github.com/home-assistant/core/issues/42793
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Change the API boundary between stream and camera

Shift more of the stream lifecycle management to the camera.  The motivation is to support stream urls that expire
giving the camera the ability to change the stream once it is created.

allenporter committed:
Document stream lifecycle and simplify stream/camera interaction

allenporter committed:
Reorder create_stream function to reduce diffs

allenporter committed:
Increase test coverage for camera_sdm.py

uvjustin commented on homeassistant/components/stream/__init__.py:
```diff
@@ -200,6 +209,14 @@ def _stop(self):
             _LOGGER.info("Stopped stream: %s", self.source)
 
 
+def _get_stream_by_source(hass, stream_source):
```
Why change the data type from a list to a dict if we have to end up looking it up?

allenporter replied:
My thinking was that stream url can no longer be the primary key since we want to support changing the url.

I thought of the new behavior as similar to how the access token check is done: on every request for a stream sequence it iterates over the streams to find the one with the matching access token.

However, it's not the only approach we could do.  Maybe also:
- change how camera invokes record, so that the lookup of existing worker sn't by stream url but by using the same stream object (considered this but was keeping diff small)
- support add/remove of the dict when the url changes
- use a different key such as camera entity id

One thing I don't have intuition for is how often the resuse of existing stream happens by calls to stream record vs calls to camera record, so I leaned towards preserving behavior as much as possible.

uvjustin replied:
I think the access token check is harder to avoid since we end up resetting the tokens periodically. For the streams itself I think it's easy to skip that iterating check by storing the streams by some other key, but as you said we will need to change the `async_handle_record_service` call to use that other key as well.

allenporter replied:
Yeah, my assumption is that what is here is better than a backwards incompatible service API call change for camera.  Even if we added an additional optional key for usage by camera, record seems like it needs to work with a raw rtsp stream url without that key, so it would still need to iterate through everything and match the rtsp stream url anyway.

uvjustin replied:
I don't think there is a problem changing the service call - it looks like it's just an internal service call between camera and stream. Instead of camera calling the service with `CONF_STREAM_SOURCE: source` we could just have camera call the service with the new key, and then on the stream side it can look up the stream directly with that key.

allenporter replied:
Are you suggesting it would be OK to remove `stream_source` from the API?

I assumed that since `stream.record` shows up in the home assistant user interface that users can build automations, etc,  to record arbitrary rtsp stream urls without a camera. (I just tried this out and it).  Just to clarify: you're saying we can break that functionality because it was never intended to be used outside of `camera`?

That is, i understand how to make the change, but don't have enough context to know that it is ok to break that feature.

uvjustin replied:
You're right, I didn't realize it was exposed there and documented as a separate service. I would be an advocate of removing that service since I don't see a use for it on its own, but that would be a breaking change and we would need an OK from some higher ups. Let's see what @hunterjm says.

allenporter replied:
OK so, given the other comment thread, we'll proceed with this breaking change.  So, then what I'll do is change the service to accept a `stream_id` of a previously created stream.  This means that `record` can no longer be invoked on a stream that does not already exist, and so the camera component will make sure the stream exists before calling record.  Thanks.

uvjustin replied:
Just to be clear, I think you mean that these changes mean that stream's `record` can no longer be invoked independently on an arbitrary url and will only called by camera, and since camera is the caller it will make sure the stream is created before calling record? I think that makes sense

allenporter replied:
Yes, ok.  Given the discussion on the main comment thread, i've changed the interface to record to take a `stream_id` rather than a `stream_source` which addresses this.  Streams can now be looked up by that id in the `ATTR_STREAMS` dict again.

allenporter replied:
Based on the other thread, the record service was removed entirely.

uvjustin commented on homeassistant/components/stream/__init__.py:
```diff
@@ -38,9 +52,12 @@
+    """Create a stream based on the source url.
+
+    The stream_source is typically an rtsp url and options are passed into
+    pyav / ffpmpeg as options.
```
small typo

uvjustin reviewed (commented):
I think @hunterjm is a better person to look at this since he's more familiar with the interaction between camera/stream. As long as it doesn't break tests I think the refactoring seems fine.

allenporter committed:
Fix ffmpeg typo.

hunterjm commented:
> I think @hunterjm is a better person to look at this since he's more familiar with the interaction between camera/stream. As long as it doesn't break tests I think the refactoring seems fine.

It was definitely intended to be an internal service. There's no telling if users are in fact using it directly, but most people I think would add their cameras to Home Assistant and use the `camera.record` service instead, so probably very low impact though technically a breaking change.

allenporter commented:
> > I think @hunterjm is a better person to look at this since he's more familiar with the interaction between camera/stream. As long as it doesn't break tests I think the refactoring seems fine.
> 
> It was definitely intended to be an internal service. There's no telling if users are in fact using it directly, but most people I think would add their cameras to Home Assistant and use the `camera.record` service instead, so probably very low impact though technically a breaking change.

OK thanks, very helpful.  @uvjustin I can use a key of the camera entity id or something then.  Thanks.

@hunterjm also looking for any other thoughts about if we're headed anywhere close to the right direction here overall.

hunterjm commented:
> @hunterjm also looking for any other thoughts about if we're headed anywhere close to the right direction here overall.

I know, and at a glance I think it looks fine. Unfortunately annual review time has kicked off at work and ultimately eats up my nights and weekends through mid-February every year because I can never find the time during the day to work on them. I don't want to be a blocker here any longer, so if @uvjustin is willing to work with you on it I'll defer to you two. At this point in time you both know just as much about PyAV and video streaming as I do.

uvjustin commented:
> 
> 
> > @hunterjm also looking for any other thoughts about if we're headed anywhere close to the right direction here overall.
> 
> I know, and at a glance I think it looks fine. Unfortunately annual review time has kicked off at work and ultimately eats up my nights and weekends through mid-February every year because I can never find the time during the day to work on them. I don't want to be a blocker here any longer, so if @uvjustin is willing to work with you on it I'll defer to you two. At this point in time you both know just as much about PyAV and video streaming as I do.

Got it. I think @allenporter and I just wanted to make sure you were ok with the changes. There's quite a lot of rearchitecting/refactoring primarily to make Nest type streams work, but I think @allenporter has mapped it out pretty well and I think his cleanup of the tests will help avoid any big problems. I'm fine with the changes as long as we keep everything working and it seems like that's your position as well.

allenporter committed:
Add a stream identifier for each stream, managed by camera

allenporter commented on homeassistant/components/camera/__init__.py:
```diff
@@ -754,7 +731,7 @@ async def async_handle_record_service(camera, call):
 
     data = {
-        CONF_STREAM_SOURCE: source,
+        CONF_STREAM_ID: stream.stream_id,
```
Another thought here is to add a method called "record" on the Stream and remove this service entirely, or move "create stream" and "update stream" to be a service rather than a class.

allenporter reviewed (commented):
I'm also happy to incorporate more feedback if we're not happy with the end result.

> > > @hunterjm also looking for any other thoughts about if we're headed anywhere close to the right direction here overall.
> > 
> > 
> > I know, and at a glance I think it looks fine. Unfortunately annual review time has kicked off at work and ultimately eats up my nights and weekends through mid-February every year because I can never find the time during the day to work on them. I don't want to be a blocker here any longer, so if @uvjustin is willing to work with you on it I'll defer to you two. At this point in time you both know just as much about PyAV and video streaming as I do.
> 
> Got it. I think @allenporter and I just wanted to make sure you were ok with the changes. There's quite a lot of rearchitecting/refactoring primarily to make Nest type streams work, but I think @allenporter has mapped it out pretty well and I think his cleanup of the tests will help avoid any big problems. I'm fine with the changes as long as we keep everything working and it seems like that's your position as well.

OK thanks.  I'm also happy to incorporate any future feedback to simplify/cleanup if we have better ideas in the future.

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -267,14 +256,10 @@ async def preload_stream(_):
                 continue
-
-            request_stream(hass, source, keepalive=True, options=camera.stream_options)
+            stream.stream_view_url("hls")
```
Isn't it weird that requesting a url has the side effect of keeping the stream alive? Should we add a new method or constructor parameter to indicate `keepalive`?

balloob replied:
Actually, I see that it is already taken care off in `create_stream`. Is this call still necessary?

allenporter replied:
I think the work happening is something like:
- `camera.create_stream` creates the Stream
- update options on the stream
- `stream.add_provider('hls')` to create the hls output and set an access token
- `stream.start` the background thread in case it died

All of this used to happen in `request_stream` so the challenge is to figure out the right division of labor into a usable API.  I was thinking the caller really is just saying: 
- Give me a stream,
- start a recording
- start an hls stream for viewing

An option here could be to add a function in camera that updates keepalive and returns the view url (I still like removing the duplicate prefs lookup).  I'll think about this more...

allenporter replied:
My takeaway here is that that this was a bit too opaque.  I've unrolled some of this logic out of the `Stream` into `camera`, which I think may make the interaction a bit more clear with fewer side effects.

balloob commented on homeassistant/components/stream/__init__.py:
```diff
@@ -1,4 +1,19 @@
+
+Components use create_stream with a stream source (e.g. an rtsp url) to create
+a new Stream object. Stream manages:
+  - Background work to fetch an decode a stream
```
```suggestion
  - Background work to fetch and decode a stream
```

balloob commented on homeassistant/components/stream/__init__.py:
```diff
@@ -38,12 +52,18 @@
-@bind_hass
-def request_stream(hass, stream_source, *, fmt="hls", keepalive=False, options=None):
-    """Set up stream with token."""
+def create_stream(hass, stream_id, stream_source, options=None):
```
Would it make sense to just generate a stream_id inside this method? 

Or actually, do we need a stream ID ? The objects are now managed by `camera`, so the camera can deal with it (it stores it on CameraEntity instance now)

allenporter replied:
We don't need it if we drop the record service (I had a similar thought above that it isn't needed so I'm on board with this)

balloob commented on homeassistant/components/stream/__init__.py:
```diff
@@ -202,7 +215,7 @@ def _stop(self):
 
 async def async_handle_record_service(hass, call):
     """Handle save video service calls."""
-    stream_source = call.data[CONF_STREAM_SOURCE]
```
This feels weird. Seems like this should be a camera service now that uses entity ID to target streams.

allenporter replied:
There is already a `camera` service that called this service.  I've removed this service entirely, and just made it a method on `Stream`.

balloob commented on homeassistant/components/stream/services.yaml:
```diff
@@ -1,6 +1,9 @@
 record:
-  description: Make a .mp4 recording from a provided stream.
+  description: Make a .mp4 recording from a provided stream. This is an internal service meant to be called by other components.
```
Internal components can call it directly on the `Stream` object that they received from calling the internal cross-integration API `create_stream`. I think that we can drop this service.

allenporter replied:
Ok great, I was proposing above we add a method called record similar to the method that gives back the url.  I'll try this out and see how it looks.

balloob reviewed (commented):
I left some comments on the architecture. It looks great!

allenporter committed:
Update homeassistant/components/stream/__init__.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

allenporter committed:
Remove stream record service

allenporter committed:
Unroll changes to Stream interface back into camera component

allenporter committed:
Fix preload stream to actually start the background worker

allenporter committed:
Reduce unncessary diffs for readability

allenporter committed:
Remove redundant camera stream start code

allenporter commented:
I've addressed all the feedback and updated the PR description to reflect the latest approach.  The scope increased a bit, but it's mostly removal of the service.  This is ready for another look.  Thanks all.

allenporter commented:
Definitely didn't intend to close this.  Whoops.

allenporter commented:
@uvjustin this is ready for re-review when you have a chance based on your feedback and the other suggestions. (sorry for the noise above, it should have settled now, so wanted to clarify).  I'm considering you the primary reviewer still if thats alright.

uvjustin commented:
> 
> 
> @uvjustin this is ready for re-review when you have a chance based on your feedback and the other suggestions. (sorry for the noise above, it should have settled now, so wanted to clarify). I'm considering you the primary reviewer still if thats alright.

Sure, I'll try to get it done in a bit.

uvjustin commented on homeassistant/components/camera/__init__.py:
```diff
@@ -375,6 +355,17 @@ def frame_interval(self):
+        """Create a Stream for stream_source."""
+        # There is at most one stream (a decode worker) per camera
+        if not self.stream:
+            async with async_timeout.timeout(10):
```
Realize this timeout value was hard coded before, but maybe we should make this a constant

allenporter replied:
I'll send a follow up to add this.

uvjustin reviewed (approved):
Looks good!
